README.md
# The Bouncy World Engine

The Bouncy World Engine is a rust-lang application that allows for defined entities to bounce around a defined world. It starts as just a black box bouncing alone in a white world but can quickly grow into something quite creative by just using the [configuration system](https://github.com/ChrWalte/Bouncy-World#configuration-system) and the [world save system](https://github.com/ChrWalte/Bouncy-World#world-save-system). Check out some of the [examples](https://github.com/ChrWalte/Bouncy-World/tree/main/examples) listed in the Source code to get some inspiration.

## Installing The Bouncy World Engine

There are a few ways to install The Bouncy World Engine, either downloading the compiled binary from the GitHub Releases section or cloning the Source Code and compiling it yourself. Both are outlined below. This application was built to run standalone, out of the box, without configuration. This is where the lonely black box comes into play. After installing, the user can configure the look and feel of both the entities bouncing around and the world itself.

### Installing from GitHub Releases

Every release under the [GitHub Releases](https://github.com/ChrWalte/Bouncy-World/releases/) section contains the current source code and the built binary. Install the binary or executable to run this application on a Windows Intel Based x64 System.

## Running The Bouncy World Engine

Running The Bouncy World Engine will start a default, lonely black box bouncing in a white world. This default configuration can be customized extensively through the [configuration system](https://github.com/ChrWalte/Bouncy-World#configuration-system) and [world save system](https://github.com/ChrWalte/Bouncy-World#world-save-system).

### Running bouncy-world.exe

Running The Bouncy World Engine via the `bouncy-world.exe` is easy. Just run the executable and watch as a default bouncy world with a single entity bouncing around.

Running The Bouncy World Engine in the current directory:

```shell
.\\bouncy-world.exe
```

### Running from the Source Code

Running The Bouncy World Engine via the Source Code requires a few things:

The source code must be downloaded locally. This can be done via [git](https://git-scm.com/), a Source Code Version Control, or directly downloaded via GitHub as a [compressed ZIP file](https://github.com/ChrWalte/Bouncy-World/archive/refs/heads/main.zip).

Cloning The Bouncy World Engine Source Code to a local directory:

```shell
git clone https://github.com/ChrWalte/Bouncy-World.git
```

The [rust toolchain](https://www.rust-lang.org/tools/install) must be installed to build and run from the source code.
Once the toolchain is installed, use `cargo` to build and run The Bouncy World Engine.

```shell
cargo run --release
```

## Command Line Interface

The Bouncy World Engine has a Command Line Interface or CLI to create configurations and world saves.

### bouncy-would.exe CLI

To see a list of available commands, run the `help` command:

```shell
.\\bouncy-world.exe help
The Bouncy World Engine - v1.0.2

Usage: bouncy-world.exe <COMMAND>

Commands:
  new           generates a new default world save (in the known-universe folder)
  save          save a copy of the current world configuration (to the known-universe folder)
  config        generates a new default world configuration (if not already present)
  help          shows a list of all commands and their description
```

### cargo run CLI

These Command Line Arguments can be used via cargo as well:

```shell
cargo run --release -- help
The Bouncy World Engine - v1.0.2

Usage: bouncy-world.exe <COMMAND>

Commands:
  new           generates a new default world save (in the known-universe folder)
  save          save a copy of the current world configuration (to the known-universe folder)
  config        generates a new default world configuration (if not already present)
  help          shows a list of all commands and their description
```

## Configuration System

The Bouncy World Engine has a configuration system that allows for bouncy worlds to be loaded from a JSON or YAML file. these files need to be in the exact format for the version you are using.

Please refer to the examples on how to create your configuration files.

### Loading from a Configuration File

Loading from a configuration file in The Bouncy World Engine is simple. Just pass in the config path that yu would like to load.

```shell
.\\bouncy-world.exe .\\known-universe\\a-ransom-save\\config.json
.\\bouncy-world.exe .\\known-universe\\a-ransom-save\\config.yaml
.\\bouncy-world.exe .\\known-universe\\a-ransom-save\\config.yml
```

### Example JSON Configuration File

Here is an example JSON configuration file:

```jsonc
{
  // the version of The Bouncy World Engine that should be used for this configuration
  "bouncy_world_engine_version": "1.0.2",
  // for logging config and world information to the console (default: false)
  "is_debug_mode": false,

  // the world dimensions (default: width: 600, height: 400)
  "world_width": 600,
  "world_height": 400,
  // the world background hex color (default: ffffffff or White)
  "world_color": "ffffffff",

  // the type of world to use (default: Color)
  "world_type": "Color",
  // example of Image world type
  // user must specify image path and width and height of image
  // "world_type": {
  //   "Image": [
  //     "path-to-image.png",
  //     50,
  //     50
  //   ]
  // },

  // the number of entities to create (default: 1)
  "entity_count": 1,
  // the color of the entities (default: 000000ff or Black)
  "entity_color": "000000ff",
  // the velocity of entities, or how fast they are moving (default: 5)
  "entity_velocity": 5,

  // the type of entities to use (Default: Box: 50, 50)
  // user must specify the Box width and height
  "entity_type": {
    "Box": [
      50,
      50
    ]
  }
  // example of Ball entity type
  // user must specify radius of Ball
  // "entity_type": {
  //   "Ball": [
  //     25
  //   ]
  // },
  // example of Image entity type
  // user must specify image path and width and height of image
  // "entity_type": {
  //   "Image": [
  //     "path-to-image.png",
  //     50,
  //     50
  //   ]
  // }
}
```

### Example YAML (or YML) Configuration File

Here is an example YAML or YML configuration file:

```yaml
# the version of The Bouncy World Engine that should be used for this configuration
bouncy_world_engine_version: 1.0.2
# for logging config and world information to the console (default: false)
is_debug_mode: false

# the world dimensions (default: width: 600, height: 400)
world_width: 600
world_height: 400
# the world background hex color (default: ffffffff or White)
world_color: ffffffff
# the type of world to use (default: Color)
world_type: Color
# example of Image world type
# user must specify image path and width and height of image
# note: the world_color is ignored if an image is used
# world_type:
#   !Image
#   - "path-to-image.png"
#   - 50
#   - 50

# the number of entities to create (default: 1)
entity_count: 1
# the color of the entities (default: 000000ff or Black)
entity_color: 000000ff
# the velocity of entities, or how fast they are moving (default: 5)
entity_velocity: 5

# the type of entities to use (Default: Box: 50, 50)
# user must specify the Box width and height
entity_type:
  !Box
  - 50
  - 50
# example of Ball entity type
# user must specify radius of Ball
# entity_type:
#   !Ball
#   - 25
# example of Image entity type
# user must specify image path and width and height of image
# note: the entity_color is ignored if an image is used
# entity_type:
#   !Image
#   - "path-to-image.png"
#   - 50
#   - 50
```

## World Save System

### Loading from a World Save File

Loading from a world save file in The Bouncy World Engine is simple. Just pass in the bouncy-world path that yu would like to load.

```shell
.\\bouncy-world.exe .\\known-universe\\a-ransom-save\\a-random-save.bouncy-world
```

### Example Default bouncy-world Save File

```yaml
# date-time-stamp of when the world save was created
date_time_stamp: 0001-01-01T01.01.01
# the version of The Bouncy World Engine that should be used for this world save
bouncy_world_engine_version: 1.0.2

# the object world itself
# the user gets a lot more control over the world and its entities
world:
  # the world dimensions
  width: 600
  height: 400

  # the world background color
  background_rgba_hex: ffffffff

  # the world type
  world_type: Color

  # example image world type
  # note: the background_rgba_hex is ignored if using image world type
  # world_type:
  #   !Image
  #   - "path-to-image.png"
  #   - 50
  #   - 50

  # the entities bouncy around in the world
  entities:
    # each entity must be listed here
    # starting x and y positions of the entity
    - x_position: 230
      y_position: 215
      # x and y velocity of the entity
      x_velocity: 5
      y_velocity: -5
      # the color of the entity
      rgba_hex: 000000ff
      # the entity type
      entity_type:
        !Box
        - 50
        - 50

      # example ball entity type
      # entity_type:
      #   !Ball
      #   - 25
      # example image entity type
      # note: the rgba_hex is ignored if using image entity type
      # entity_type:
      #   !Image
      #   - "path-to-image.png"
      #   - 50
      #   - 50
```


CHANGELOG.md
# The Bouncy World Engine Change Log

## Version 1.0.3

- dependency updates

## Version 1.0.2

- dependency updates

## Version 1.0.1

- dependency updates

## Version 1.0.0

- Initial release!
- Created base of writing pixels to screen based on World and Entity colors.
- Entities bounce around World Object based on positioning.
- Command Line Interface to create and load World Save Files and Configuration Files.


src/cli.rs
use std::path::Path;

use crate::{
    config::Config,
    constants::{
        CONFIG_COMMAND_DESCRIPTION, CONFIG_COMMAND_LONG, CONFIG_COMMAND_SHORT,
        HELP_COMMAND_DESCRIPTION, HELP_COMMAND_LONG, HELP_COMMAND_SHORT, JSON_CONFIG_PATH,
        NEW_COMMAND_DESCRIPTION, NEW_COMMAND_LONG, NEW_COMMAND_SHORT, SAVE_COMMAND_DESCRIPTION,
        SAVE_COMMAND_LONG, SAVE_COMMAND_SHORT, WORLD_SAVE_FILE_EXTENSION, YAML_CONFIG_PATH,
        YML_CONFIG_PATH,
    },
    save::Save,
    world::World,
};

#[derive(Eq, PartialEq, Debug)]
pub enum BouncyWorldWindow {
    Show,
    Hide,
}

// TODO: add version command
// TODO: add update command
pub fn run_command(given_command: &str) -> (Config, Save, BouncyWorldWindow) {
    // start with default config & an empty world save
    let mut running_config: Config = Config::new();
    let running_world_save: Save;

    if Path::new(&given_command).exists() {
        // the argument is a valid path, try loading it as a world save or config
        (running_config, running_world_save) = try_load_from_files(&given_command);
        return (running_config, running_world_save, BouncyWorldWindow::Show);
    } else if given_command == NEW_COMMAND_LONG || given_command == NEW_COMMAND_SHORT {
        // run new command:
        running_config = Config::new();
        running_world_save = Save::new(World::new(&running_config));
        let save_file_location = running_world_save.save(&running_config);
        println!(
            "saved default config and default world save to known universe: {}",
            save_file_location
        );
    } else if given_command == SAVE_COMMAND_LONG || given_command == SAVE_COMMAND_SHORT {
        // run save command:
        running_config = if Path::new(YAML_CONFIG_PATH).exists() {
            Config::load_from_yaml_file(YAML_CONFIG_PATH)
        } else if Path::new(YML_CONFIG_PATH).exists() {
            Config::load_from_yaml_file(YML_CONFIG_PATH)
        } else if Path::new(JSON_CONFIG_PATH).exists() {
            Config::load_from_json_file(JSON_CONFIG_PATH)
        } else {
            println!("no config found, loaded default config");
            running_config
        };

        running_world_save = Save::new(World::new(&running_config));
        let save_file_location = running_world_save.save(&running_config);
        println!(
            "saved config and world save to known universe: {}",
            save_file_location
        );
    } else if given_command == CONFIG_COMMAND_LONG || given_command == CONFIG_COMMAND_SHORT {
        // run config command:
        if !Path::new(JSON_CONFIG_PATH).exists() {
            running_config = Config::new();
            // TODO: add ability to save to yaml file
            // running_config.save_to_yaml_file(JSON_CONFIG_PATH);
            running_config.save_to_json_file(JSON_CONFIG_PATH);
            println!("save config file in current directory");
        } else {
            println!("config file already exists in current directory");
        }
        std::process::exit(0);
    } else if given_command == HELP_COMMAND_LONG || given_command == HELP_COMMAND_SHORT {
        // run help command:
        print_help_message();
        std::process::exit(0);
    } else {
        println!("unknown command: {:?}", given_command);
        std::process::exit(0);
    }

    (running_config, running_world_save, BouncyWorldWindow::Hide)
}

fn try_load_from_files(given_path: &str) -> (Config, Save) {
    // start with an empty config & an empty world save
    let running_config: Config;
    let running_world_save: Save;

    let path = Path::new(&given_path);
    let path_file_extension = path
        .extension()
        .expect("expected config file or bouncy-world file");
    if path_file_extension == WORLD_SAVE_FILE_EXTENSION {
        (running_config, running_world_save) = try_load_from_bouncy_world_save(path);
    } else if path.ends_with(YAML_CONFIG_PATH) || path.ends_with(YML_CONFIG_PATH) {
        (running_config, running_world_save) = load_config_from_yaml(given_path);
    } else if path.ends_with(JSON_CONFIG_PATH) {
        (running_config, running_world_save) = load_config_from_json(given_path);
    } else {
        println!("unknown command: {:?}", given_path);
        std::process::exit(0);
    }

    (running_config, running_world_save)
}

fn try_load_from_bouncy_world_save(world_save_path: &Path) -> (Config, Save) {
    // start with default config & default world save
    let mut running_config: Config = Config::new();
    let mut running_world_save: Save = Save::new(World::new(&running_config));

    // try to load world, use default if fails
    running_world_save = match Save::load(world_save_path) {
        Some(world_save) => {
            println!("loaded from world save");
            world_save
        }
        None => {
            println!("unable to load save, loaded default world save");
            running_world_save
        }
    };

    // read from config in same directory as the world save
    let world_save_dir = world_save_path
        .parent()
        .expect("could not get directory from path");
    let world_save_dir = world_save_dir
        .to_str()
        .expect("could not convert path to a str");
    let world_yaml_config = &format!("{}\\{}", world_save_dir, YAML_CONFIG_PATH);
    let world_yml_config = &format!("{}\\{}", world_save_dir, YML_CONFIG_PATH);
    let world_json_config = &format!("{}\\{}", world_save_dir, JSON_CONFIG_PATH);

    running_config = if Path::new(world_yaml_config).exists() {
        Config::load_from_yaml_file(world_yaml_config)
    } else if Path::new(world_yml_config).exists() {
        Config::load_from_yaml_file(world_yml_config)
    } else if Path::new(world_json_config).exists() {
        Config::load_from_json_file(world_json_config)
    }
    // if no config was found, try to load config from current directory
    else if Path::new(YAML_CONFIG_PATH).exists() {
        Config::load_from_yaml_file(YAML_CONFIG_PATH)
    } else if Path::new(YML_CONFIG_PATH).exists() {
        Config::load_from_yaml_file(YML_CONFIG_PATH)
    } else if Path::new(JSON_CONFIG_PATH).exists() {
        Config::load_from_json_file(JSON_CONFIG_PATH)
    } else {
        println!("no config found, created config from world save");
        Config::reverse_from_world_save(&running_world_save)
    };

    (running_config, running_world_save)
}

fn load_config_from_yaml(config_path: &str) -> (Config, Save) {
    let running_config = Config::load_from_yaml_file(config_path);
    let running_world_save = Save::new(World::new(&running_config));
    println!("loaded from yaml config");

    (running_config, running_world_save)
}

fn load_config_from_json(config_path: &str) -> (Config, Save) {
    let running_config = Config::load_from_json_file(config_path);
    let running_world_save = Save::new(World::new(&running_config));
    println!("loaded from json config");

    (running_config, running_world_save)
}

fn print_help_message() {
    println!("The Bouncy World Engine - v{}\n", env!("CARGO_PKG_VERSION"));
    println!("Usage: bouncy-world.exe <COMMAND>\n");
    println!("Commands:");
    println!("  {}\t\t{}", NEW_COMMAND_LONG, NEW_COMMAND_DESCRIPTION);
    println!("  {}\t\t{}", SAVE_COMMAND_LONG, SAVE_COMMAND_DESCRIPTION);
    println!("  {}\t{}", CONFIG_COMMAND_LONG, CONFIG_COMMAND_DESCRIPTION);
    println!("  {}\t\t{}", HELP_COMMAND_LONG, HELP_COMMAND_DESCRIPTION);
}


src/main.rs
use std::{collections::HashMap, env, path::Path};

use cli::BouncyWorldWindow;
use config::Config;
use constants::{JSON_CONFIG_PATH, YAML_CONFIG_PATH, YML_CONFIG_PATH};
use entity::EntityType;
use fltk::{
    app,
    prelude::{GroupExt, WidgetExt},
    window::Window,
};
use pixels::{Pixels, SurfaceTexture};
use save::Save;
use world::{World, WorldType};

pub mod cli;
pub mod config;
pub mod constants;
pub mod entity;
pub mod save;
pub mod world;

// TODO: ability to pause application loop in debug mode
// TODO: ability to take control of entity in debug mode
// TODO: ability to save images of pixel frames
// TODO: ability to save video of bouncy-world

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() > 2 {
        println!("unknown command(s): {:?}", args);
        return;
    }

    // start with default config & empty world save
    let mut running_config: Config = Config::new();
    let mut running_world_save: Save;

    // condition to show bouncy-world window
    let mut show_bouncy_world_window = BouncyWorldWindow::Show;

    if args.len() == 2 {
        // one argument was passed in from user
        let given_command = &args[1];
        (running_config, running_world_save, show_bouncy_world_window) =
            cli::run_command(given_command);
    } else {
        // no arguments passed in, try to load config from current directory
        running_config = if Path::new(YAML_CONFIG_PATH).exists() {
            Config::load_from_yaml_file(YAML_CONFIG_PATH)
        } else if Path::new(YML_CONFIG_PATH).exists() {
            Config::load_from_yaml_file(YML_CONFIG_PATH)
        } else if Path::new(JSON_CONFIG_PATH).exists() {
            Config::load_from_json_file(JSON_CONFIG_PATH)
        } else {
            // no config found, use default config
            running_config
        };

        let running_world = World::new(&running_config);
        running_world_save = Save::new(running_world);
    }

    if running_config.is_debug_mode {
        dbg!(&running_config, &running_world_save);
    }

    if show_bouncy_world_window == BouncyWorldWindow::Show {
        // initialize fltk app and fltk window
        let app = app::App::default();
        let mut window = Window::default()
            .with_size(running_config.world_width, running_config.world_height)
            .with_label("Bouncy World");
        window.end();
        window.show();

        // load images into memory
        let mut running_images = HashMap::new();
        for entity in &running_world_save.world.entities {
            match &entity.entity_type {
                EntityType::Image(image_path, _, _) => {
                    running_images.insert(
                        image_path.to_string(),
                        image::open(image_path).expect("could not read image at given path"),
                    );
                }
                _ => (), // not an image
            };
        }
        match &running_world_save.world.world_type {
            WorldType::Image(image_path, _, _) => {
                running_images.insert(
                    image_path.to_string(),
                    image::open(image_path).expect("could not read image at given path"),
                );
            }
            _ => (), // not an image
        }

        // initialize pixels
        let mut pixels = {
            let pixel_width = window.pixel_w() as u32;
            let pixel_height = window.pixel_h() as u32;
            let surface_texture = SurfaceTexture::new(pixel_width, pixel_height, &window);
            Pixels::new(
                running_world_save.world.width as u32,
                running_world_save.world.height as u32,
                surface_texture,
            )
            .expect("pixels failed to initialize")
        };

        // fltk app loop
        while app.wait() {
            // handle events

            // update internal world state
            running_world_save.world.update(&running_config);

            // draw the current frame
            running_world_save
                .world
                .draw(pixels.frame_mut(), &running_config, &running_images);
            pixels.render().expect("pixels failed to render");

            // redraw window and trigger event loop
            app::flush();
            app::awake();
        }
    }
}


src/save.rs
use std::{fs, path::Path};

use serde_derive::{Deserialize, Serialize};

use crate::{
    config::Config,
    constants::{WORLD_SAVE_FILE_EXTENSION, WORLD_SAVE_LOCATION, YAML_CONFIG_PATH},
    world::World,
};

#[derive(Deserialize, Serialize, Debug)]
pub struct Save {
    pub world: World,
    pub date_time_stamp: String,
    pub bouncy_world_engine_version: String,
}

impl Save {
    pub fn new(world: World) -> Save {
        Save {
            world,
            // TODO: add am/pm in format
            date_time_stamp: chrono::offset::Local::now()
                .format("%Y-%m-%dT%H.%M.%S")
                .to_string(),
            bouncy_world_engine_version: env!("CARGO_PKG_VERSION").to_string(),
        }
    }

    pub fn load(save_path: &Path) -> Option<Save> {
        // TODO: ability to load config via json?
        if save_path.exists() {
            let save_str = fs::read_to_string(save_path).expect("could not read from file");
            Some(serde_yaml::from_str::<Save>(&save_str).expect("could not read as a save file"))
        } else {
            None
        }
    }

    pub fn save(&self, config: &Config) -> String {
        fs::create_dir_all(WORLD_SAVE_LOCATION)
            .expect("could not create dir at WORLD_SAVE_LOCATION");

        let save_path = &format!("{}\\{}", WORLD_SAVE_LOCATION, &self.date_time_stamp);
        fs::create_dir_all(save_path).expect("could not create dir at self.date_time_stamp");

        let save_str = serde_yaml::to_string(&self).expect("could not write config as yaml");

        std::fs::write(
            format!(
                "{}\\{}.{}",
                save_path, &self.date_time_stamp, WORLD_SAVE_FILE_EXTENSION
            ),
            save_str,
        )
        .expect("could not write to file");

        let config_file_location = format!("{}\\{}", save_path, YAML_CONFIG_PATH);
        config.save_to_yaml_file(&config_file_location);

        save_path.to_string()
    }
}


src/world.rs
use std::collections::HashMap;

use crate::{
    config::Config,
    entity::{Entity, EntityType},
};
use hex::FromHex;
use image::{DynamicImage, GenericImageView};
use serde_derive::{Deserialize, Serialize};

#[derive(Eq, PartialEq, Hash, Deserialize, Serialize, Debug)]
pub struct World {
    pub width: i32,
    pub height: i32,
    pub world_type: WorldType,
    pub background_rgba_hex: String,
    pub entities: Vec<Entity>,
}

#[derive(Clone, Eq, PartialEq, Hash, Deserialize, Serialize, Debug)]
pub enum WorldType {
    Color,
    // TODO: change this to only the image path and get the width and height dynamically.
    Image(String, i32, i32),
}

fn get_hex_pixel_color_from_image(
    pixel_x_position: i32,
    pixel_y_position: i32,
    image_path: &String,
    images: &HashMap<String, DynamicImage>,
) -> String {
    let image = images
        .get(image_path)
        .expect("image_path not found in images");

    let rgba = image
        .get_pixel((pixel_x_position) as u32, (pixel_y_position) as u32)
        .0;
    let [r, g, b, a] = rgba;
    format!("{:02x}{:02x}{:02x}{:02x}", r, g, b, a)
}

impl World {
    pub fn new(config: &Config) -> World {
        let mut world = World {
            width: config.world_width,
            height: config.world_height,
            world_type: config.world_type.clone(),
            background_rgba_hex: config.world_color.to_string(),
            entities: vec![],
        };

        for _ in (0..config.entity_count).into_iter() {
            world.entities.push(Entity::new(&config));
        }

        world
    }

    // update world every frame
    pub fn update(&mut self, config: &Config) {
        // loop through all entities and update each
        for entity in &mut self.entities {
            entity.update(&config);
        }
    }

    // draw world every frame, the world handles the color of each pixel
    pub fn draw(
        &mut self,
        frame: &mut [u8],
        config: &Config,
        images: &HashMap<String, DynamicImage>,
    ) {
        let mut image_entity_starting_positions: HashMap<usize, (i32, i32)> = HashMap::new();

        // loop through each pixel (frame split in four due to rrggbbaa format)
        for (i, pixel) in frame.chunks_exact_mut(4).enumerate() {
            // calculate pixels x and y positions on frame
            let pixel_x_position = (i % config.world_width as usize) as i32;
            let pixel_y_position = (i / config.world_width as usize) as i32;

            // TODO: remove?
            let mut temp_entity_rgba_hex_str;

            let mut is_pixel_occupied = false;
            let mut rgba_hex_str = match &self.world_type {
                WorldType::Color => &self.background_rgba_hex,
                WorldType::Image(image_path, width, height) => {
                    if pixel_x_position <= *width && pixel_y_position <= *height {
                        temp_entity_rgba_hex_str = get_hex_pixel_color_from_image(
                            pixel_x_position,
                            pixel_y_position,
                            image_path,
                            images,
                        );

                        &temp_entity_rgba_hex_str
                    } else {
                        &self.background_rgba_hex
                    }
                }
            };
            for (entity_index, entity) in self.entities.iter().enumerate() {
                if entity.is_within_entity(pixel_x_position, pixel_y_position) {
                    rgba_hex_str = match &entity.entity_type {
                        EntityType::Box(_, _) => &entity.rgba_hex,
                        EntityType::Ball(_) => &entity.rgba_hex,
                        EntityType::Image(image_path, _, _) => {
                            if !image_entity_starting_positions.contains_key(&entity_index) {
                                image_entity_starting_positions
                                    .insert(entity_index, (pixel_x_position, pixel_y_position));
                            }

                            if !is_pixel_occupied {
                                is_pixel_occupied = true;
                                let (starting_image_x_position, starting_image_y_position) =
                                image_entity_starting_positions.get(&entity_index).expect(
                                    "failed to get image_entity_starting_positions by entity_index",
                                );

                                let entity_hex_color = get_hex_pixel_color_from_image(
                                    pixel_x_position - starting_image_x_position,
                                    pixel_y_position - starting_image_y_position,
                                    image_path,
                                    images,
                                );

                                if &entity_hex_color[6..] == "00" {
                                    is_pixel_occupied = false;
                                    &rgba_hex_str
                                } else {
                                    temp_entity_rgba_hex_str = entity_hex_color;
                                    &temp_entity_rgba_hex_str
                                }
                            } else {
                                &rgba_hex_str
                            }
                        }
                        EntityType::Text(_, _) => &entity.rgba_hex,
                    };
                }
            }

            pixel.copy_from_slice(
                &<[u8; 4]>::from_hex(rgba_hex_str).expect(
                    "issue converting rgba_str to rgba u8 slice, expected format: rrggbbaa",
                ),
            );
        }
    }
}


src/config.rs
use std::fs;

use serde_derive::{Deserialize, Serialize};

use crate::{
    constants::{
        DEBUG_MODE, DEFAULT_ENTITY_COLOR, DEFAULT_ENTITY_COUNT, DEFAULT_ENTITY_TYPE,
        DEFAULT_ENTITY_VELOCITY, DEFAULT_WORLD_COLOR, DEFAULT_WORLD_HEIGHT, DEFAULT_WORLD_TYPE,
        DEFAULT_WORLD_WIDTH,
    },
    entity::EntityType,
    save::Save,
    world::WorldType,
};

#[derive(Deserialize, Serialize, Debug)]
pub struct Config {
    pub is_debug_mode: bool,

    pub world_width: i32,
    pub world_height: i32,
    pub world_color: String,
    pub world_type: WorldType,

    pub entity_count: i32,
    pub entity_color: String,
    pub entity_type: EntityType,
    pub entity_velocity: i32,

    pub bouncy_world_engine_version: String,
}

impl Config {
    pub fn new() -> Config {
        Config {
            is_debug_mode: DEBUG_MODE,
            world_width: DEFAULT_WORLD_WIDTH,
            world_height: DEFAULT_WORLD_HEIGHT,
            world_color: DEFAULT_WORLD_COLOR.to_string(),
            world_type: DEFAULT_WORLD_TYPE,
            entity_count: DEFAULT_ENTITY_COUNT,
            entity_color: DEFAULT_ENTITY_COLOR.to_string(),
            entity_velocity: DEFAULT_ENTITY_VELOCITY,
            entity_type: DEFAULT_ENTITY_TYPE,

            bouncy_world_engine_version: env!("CARGO_PKG_VERSION").to_string(),
        }
    }

    pub fn reverse_from_world_save(world_save: &Save) -> Config {
        Config {
            // use default debug mode:
            is_debug_mode: DEBUG_MODE,

            // use world settings from world save:
            world_width: world_save.world.width,
            world_height: world_save.world.height,
            world_color: world_save.world.background_rgba_hex.to_string(),
            world_type: world_save.world.world_type.clone(),

            // only the entity count can be gotten from the world save
            entity_count: world_save.world.entities.len() as i32,
            entity_color: DEFAULT_ENTITY_COLOR.to_string(),
            entity_velocity: DEFAULT_ENTITY_VELOCITY,
            entity_type: DEFAULT_ENTITY_TYPE,

            bouncy_world_engine_version: world_save.bouncy_world_engine_version.to_string(),
        }
    }

    pub fn load_from_json_file(config_json_path: &str) -> Config {
        let config_str = fs::read_to_string(config_json_path).expect("could not read from file");
        serde_json::from_str::<Config>(&config_str).expect("could not read config as json")
    }

    pub fn save_to_json_file(&self, config_json_path: &str) {
        let config_str =
            serde_json::to_string_pretty(&self).expect("could not write config as json");
        std::fs::write(config_json_path, config_str).expect("could not write to file");
    }

    pub fn load_from_yaml_file(config_yaml_path: &str) -> Config {
        let config_str = fs::read_to_string(config_yaml_path).expect("could not read from file");
        serde_yaml::from_str::<Config>(&config_str).expect("could not read config as yaml")
    }

    pub fn save_to_yaml_file(&self, config_yaml_path: &str) {
        let config_str = serde_yaml::to_string(&self).expect("could not write config as yaml");
        std::fs::write(config_yaml_path, config_str).expect("could not write to file");
    }
}


src/constants.rs
use crate::{entity::EntityType, world::WorldType};

// cli commands:
// new command:
pub const NEW_COMMAND_LONG: &str = "new";
pub const NEW_COMMAND_SHORT: &str = "n";
pub const NEW_COMMAND_DESCRIPTION: &str =
    "generates a new default world save (in the known-universe folder)";
// save command:
pub const SAVE_COMMAND_LONG: &str = "save";
pub const SAVE_COMMAND_SHORT: &str = "s";
pub const SAVE_COMMAND_DESCRIPTION: &str =
    "save a copy of the current world configuration (to the known-universe folder)";
// config command:
pub const CONFIG_COMMAND_LONG: &str = "config";
pub const CONFIG_COMMAND_SHORT: &str = "c";
pub const CONFIG_COMMAND_DESCRIPTION: &str =
    "generates a new default world configuration (if not already present)";
// help command:
pub const HELP_COMMAND_LONG: &str = "help";
pub const HELP_COMMAND_SHORT: &str = "h";
pub const HELP_COMMAND_DESCRIPTION: &str = "shows a list of all commands and their description";

// configuration files:
pub const YAML_CONFIG_PATH: &str = "config.yaml";
pub const YML_CONFIG_PATH: &str = "config.yml";
pub const JSON_CONFIG_PATH: &str = "config.json";

// bouncy-world save location and file extension
pub const WORLD_SAVE_LOCATION: &str = "known-universe";
pub const WORLD_SAVE_FILE_EXTENSION: &str = "bouncy-world";
// pub const IMAGE_SAVE_LOCATION: &str = "images";
// pub const VIDEO_SAVE_LOCATION: &str = "videos";

// configuration defaults:
// used to log some useful debug information
pub const DEBUG_MODE: bool = false;
// default world settings
pub const DEFAULT_WORLD_WIDTH: i32 = 600;
pub const DEFAULT_WORLD_HEIGHT: i32 = 400;
pub const DEFAULT_WORLD_COLOR: &str = "ffffffff";
pub const DEFAULT_WORLD_TYPE: WorldType = WorldType::Color;
// default entity settings
pub const DEFAULT_ENTITY_COUNT: i32 = 1;
pub const DEFAULT_ENTITY_VELOCITY: i32 = 5;
pub const DEFAULT_ENTITY_COLOR: &str = "000000ff";
pub const DEFAULT_ENTITY_TYPE: EntityType = EntityType::Box(50, 50);
